clippy_dev/src/sync.rs

```diff
@@ -0,0 +1,230 @@
+use std::fmt::Write;
+use std::path::Path;
+use std::process;
+use std::process::exit;
+
+use chrono::offset::Utc;
+use xshell::{cmd, Shell};
+
+use crate::{clippy_project_root, update_lints};
+
+const JOSH_FILTER: &str = ":rev(20b085d500dfba5afe0869707bf357af3afe20be:prefix=src/tools/clippy):/src/tools/clippy";
+const JOSH_PORT: &str = "42042";
+
+fn start_josh() -> impl Drop {
+    // Create a wrapper that stops it on drop.
+    struct Josh(process::Child);
+    impl Drop for Josh {
+        fn drop(&mut self) {
+            #[cfg(unix)]
+            {
+                // Try to gracefully shut it down.
+                process::Command::new("kill")
+                    .args(["-s", "INT", &self.0.id().to_string()])
+                    .output()
+                    .expect("failed to SIGINT josh-proxy");
+                // Sadly there is no "wait with timeout"... so we just give it some time to finish.
+                std::thread::sleep(std::time::Duration::from_secs(1));
+                // Now hopefully it is gone.
+                if self.0.try_wait().expect("failed to wait for josh-proxy").is_some() {
+                    return;
+                }
+            }
+            // If that didn't work (or we're not on Unix), kill it hard.
+            eprintln!("I have to kill josh-proxy the hard way, let's hope this does not break anything.");
+            self.0.kill().expect("failed to SIGKILL josh-proxy");
+        }
+    }
+
+    // Determine cache directory.
+    let local_dir = {
+        let user_dirs = directories::ProjectDirs::from("org", "rust-lang", "clippy-josh").unwrap();
+        user_dirs.cache_dir().to_owned()
+    };
+    println!("Using local cache directory: {}", local_dir.display());
+
+    // Start josh, silencing its output.
+    let mut cmd = process::Command::new("josh-proxy");
+    cmd.arg("--local").arg(local_dir);
+    cmd.arg("--remote").arg("https://github.com");
+    cmd.arg("--port").arg(JOSH_PORT);
+    cmd.arg("--no-background");
+    cmd.stdout(process::Stdio::null());
+    cmd.stderr(process::Stdio::null());
+    let josh = cmd
+        .spawn()
+        .expect("failed to start josh-proxy, make sure it is installed");
+    // Give it some time so hopefully the port is open.
+    std::thread::sleep(std::time::Duration::from_secs(1));
+
+    Josh(josh)
+}
+
+fn rustc_hash() -> String {
+    let sh = Shell::new().expect("failed to create shell");
+    sh.set_var("RUSTUP_TOOLCHAIN", "");
```

Comment:
```suggestion
    // Make sure we pick up the updated toolchain (usually rustup pins the toolchain
    // inside a single cargo/rustc invocation via this env var).
    sh.set_var("RUSTUP_TOOLCHAIN", "");
```

Reply:
Thanks! Now that everything is working, I will go through that code again and clean everything up.